WIP: finish merging Kinto and Cliquet

glasserc opened the pull request:
This is a prototype of what it might look like to merge Cliquet into Kinto. The actual merge itself is  d1e4fecb42b6ccc311722a71d29dfa097f0aaf87, which resolves conflicts in "shared" files like `requirements.txt`, `Makefile`, `setup.py`. This PR is about addressing the aftermath -- moving the `cliquet` package to `kinto.core`, moving the `cliquet_docs` into somewhere in the `docs` tree, and removing some of the duplicated/unneeded code now that the two projects are again one.

Overall highlights:
- all `cliquet.foo` settings are now `kinto.foo` settings
- `cliquet` became `kinto/core`, except for `cliquet/tests` which became `kinto/tests/core`. This is necessary because everything in `kinto` has to have 100% coverage, which shouldn't be necessary for `kinto/core/tests`.
- get rid of `kinto/core/scripts/cliquet.py`; keep only the `migrate` function, which now lives in `kinto/core/scripts.py`
- `cliquet_docs` is now `docs/core` and it is no longer copied to Kinto docs
- break backwards compatibility in a couple of important ways: no longer do anything fancy with "prefixes" in settings, and delete `initialize_cliquet`

There are still references in this project to cliquet_fxa and cliquet_pusher, which I considered outside the scope of this PR, but I'm interested in ideas on how to deal with this without breaking master.

This PR is against the `merge-from-cliquet` branch, which may eventually hit master, but in the meantime allows us to explore what it would look like to merge the cliquet and kinto repositories without breaking master (thanks @leplatrem for the idea).

glasserc committed:
git mv cliquet kinto/core

All the tests are broken, of course.

glasserc committed:
code-level rename

This changes all the imports, strings that correspond to imports, and
setting names.

This does NOT touch the backwards-compatibility hack in the settings
code that tries to backfill cliquet.setting with setting.

glasserc committed:
update URLs to point to kinto instead of cliquet

glasserc committed:
update loadtest settings

glasserc committed:
fix test_events

pytest assumes that testapp is a test. Rename the function to
make_testapp to get around this.

glasserc committed:
move tests from kinto/core to kinto/tests/core

Anything in kinto/ that isn't in kinto/tests/ has to have 100%
coverage. That doesn't seem valuable to me.

At the same time, we want to maintain a distinction between the "core"
part of the library which people might build servers out of, and the
non-core thing that people will just run.

kinto/tests/core seems like a good compromise.

glasserc committed:
kinto.core.tests -> kinto.tests.core in code

glasserc committed:
rename cliquet_ imports to core_

glasserc committed:
get rid of kinto/core/scripts/cliquet

There's really only one thing here that's valuable, and that's the
actual migrate command. Leave that around in case someone else wants
it. The fact that kinto, the major cliquet application to date, doesn't
use anything else here, suggests we don't need it. Perhaps we were
keeping it around for the sake of having an example for other
authors. In that case, the kinto script is probably sufficient.

glasserc committed:
get rid of initialize_cliquet

We're going to have to release a major version anyhow. May as well get
rid of this deprecated code.

glasserc committed:
get rid of cliquet_protocol_version

glasserc committed:
get rid of public settings prefixes

glasserc committed:
update config section in prod docs

We no longer need both; the settings are the same, and everything that
was `cliquet` is now under `kinto`.

glasserc committed:
update docker stuff

glasserc committed:
search-and-replace in cliquet_docs

The name of the thing formerly known as Cliquet is now Kinto-Core. This
isn't super ideal and another cleanup pass will be necessary on this
documentation to make sure it makes sense.

cliquet_fxa is still a real project so we maintain references to it.

This doesn't touch any of the Sphinx configuration, which we'll have to
merge with the Kinto doc config.

glasserc committed:
rename cliquet to kinto in docs

This commit covers all the "no-brainer" cases, like import names and
config parameters.

glasserc committed:
reflow tables

glasserc committed:
update settings docs to stop pointing at Cliquet

glasserc committed:
replace description of Cliquet in FAQ

glasserc committed:
update name of Kinto-Core project in source

glasserc committed:
remove reference to Cliquet "resources" in concepts

glasserc committed:
move cliquet_docs to docs/core

glasserc committed:
no longer need to futz around with Cliquet docs

glasserc committed:
merge cliquet docs/conf.py into kinto docs/conf.py

glasserc committed:
bump version in conf.py

glasserc committed:
delete old cliquet docs/conf.py

Natim commented:
- Apparently postgresql tests aren't skipped anymore if postgresql backend dependencies are missing

Natim commented:
Do you mind activating a travis build on your fork?

Natim commented:
Some flake8 errors:

```
flake8 runtests: commands[0] | flake8 kinto
kinto/__init__.py:43:24: E128 continuation line under-indented for visual indent
kinto/__init__.py:44:24: E128 continuation line under-indented for visual indent
kinto/tests/core/test_cache.py:10:28: E128 continuation line under-indented for visual indent
kinto/tests/core/test_cache.py:11:28: E128 continuation line under-indented for visual indent
kinto/tests/core/test_initialization.py:74:28: E128 continuation line under-indented for visual indent
kinto/tests/core/test_initialization.py:86:28: E128 continuation line under-indented for visual indent
kinto/tests/core/test_permission.py:9:33: E128 continuation line under-indented for visual indent
kinto/tests/core/test_permission.py:10:33: E128 continuation line under-indented for visual indent
kinto/tests/core/test_permission.py:112:80: E501 line too long (81 > 79 characters)
kinto/tests/core/resource/test_events.py:9:29: E128 continuation line under-indented for visual indent
kinto/core/initialization.py:476:80: E501 line too long (80 > 79 characters)
kinto/core/storage/__init__.py:231:80: E501 line too long (82 > 79 characters)
kinto/core/resource/model.py:78:80: E501 line too long (82 > 79 characters)
```

Natim commented:
We should deactivate the obligation of a 100% coverage for py27-raw because a lot of tests are skipped.

Natim commented on Dockerfile:
```diff
@@ -11,7 +11,6 @@ RUN \
     apt-get install -y python3 python3-setuptools python3-pip libpq5; \
     apt-get install -y build-essential git python3-dev libssl-dev libffi-dev libpq-dev; \
-    pip3 install cliquet[postgresql,monitoring]; \
     pip3 install -e /code; \
```
`pip3 install -e /code[postgresql,monitoring]`

glasserc committed:
disable coverage for tests with py2.7-raw

Thanks @Natim for suggesting this.

glasserc committed:
put back postgres and monitoring in the dockerfile

Thanks @Natim for catching this. I knew it was wrong once but then I forgot.

glasserc committed:
reindent stuff to fix flake8

glasserc committed:
reflow everything to fix other flake8 errors

glasserc commented:
OK, I activated Travis on this fork. I think there are still some outstanding test failures, but I've addressed all your other comments so far.

glasserc commented:
Oh, but it seems like postgres tests _are_ getting skipped on my machine during the py2.7 tests, so I'm not sure what you're seeing that's different.

Natim commented:
> it seems like postgres tests are getting skipped on my machine during the py2.7 tests

Yes I think it was because `SQLAlchemy` was installed but not `psycopg2`

glasserc committed:
anticipate renaming cliquet-fxa to kinto-fxa

glasserc committed:
oops, missed a reference to cliquet

glasserc committed:
remove reference to cliquet_docs

glasserc committed:
fix typo "excuted"

glasserc committed:
grammar: "is run"

leplatrem commented on kinto/tests/core/test_initialization.py:
```diff
@@ -166,74 +152,74 @@ def test_does_raise_valueerror_if_entries_are_not_hashable(self):
+        #     'kinto.paginate_by': 3.14,
+        # }
+        # with self.assertRaises(ValueError):
+        #     self.settings(settings)
```
desired ?

glasserc replied:
I left this in because I didn't know what I was going to do with the settings prefix stuff. I've removed them from public settings, but haven't taken them out of this yet. I'm going to delete this particular block of code but leave the remainder of the test, which checks that `kinto.setting` is the same as `setting`. Good catch!

leplatrem commented on CHANGELOG.rst:
```diff
@@ -9,6 +9,20 @@ This document describes changes between each past release.
   Addresses #687.
+- Removed ``initialize_cliquet()``, which has been deprecated for a while.
+- Removed ``cliquet_protocol_version``. Kinto already defines API
+  variations as part of its URL format (e.g. ``/v0``,
```
nit: «breaking API variations» ?

leplatrem commented on docs/concepts.rst:
```diff
@@ -110,13 +110,10 @@ See also
+*Kinto* relies on :rtd:`Cornice <cornice>` and :rtd:`Pyramid <pyramid>`
 to implement the endpoints and HTTP request/response handling.
 
 .. image:: images/architecture.svg
```
This file should probably be modified :)

leplatrem commented:
I see that there are some `«Kinto-Core»` left that don't always make sense. Do we leave them and tackle this later?

leplatrem commented:
I think we're good! impressive work Ethan, congratz

glasserc committed:
clarify that the API variations are major versions

Thanks @leplatrem for the suggestion.

glasserc committed:
remove commented-out test

We removed prefixes from the "hello" view, which means this assertion
is *definitely* meaningless. As for the prefixes on incoming settings,
let's leave that as it's outside the scope of this PR.

glasserc commented:
I didn't want to go too crazy on rewriting docs, figuring that this PR would already be quite large. I'm happy to do it either as part of this PR or as a follow-up.

almet commented on kinto/authorization.py:
```diff
@@ -139,10 +139,10 @@ def build_permissions_set(object_uri, unbound_permission,
 
 @implementer(IAuthorizationPolicy)
-class AuthorizationPolicy(cliquet_authorization.AuthorizationPolicy):
+class AuthorizationPolicy(core_authorization.AuthorizationPolicy):
```
`from kinto import core`; then use `core.authorization` elsewhere?

glasserc replied:
I like this idea, but in order to make it work, I have to change it to

``` python
import kinto.core.authorization
from kinto import core
```

... because `core.authorization` is only defined as a side-effect of importing `kinto.core.authorization`. This is an unfortunate quirk of Python's import behavior. I don't think this is an improvement, and `flake8` doesn't like it either.

almet replied:
Gotcha, it's a nitpick anyway!

almet commented on kinto/config/__init__.py:
```diff
@@ -2,7 +2,7 @@
 from time import strftime
 
-from kinto.core import utils as cliquet_utils
+from kinto.core import utils as core_utils
```
same remark about the imports

almet commented on kinto/tests/test_views_flush.py:
```diff
@@ -1,7 +1,7 @@
 from pyramid.config import Configurator
 
-from kinto.tests.core import support as cliquet_support
+from kinto.tests.core import support as core_support
```
ditto

almet commented on kinto/tests/support.py:
```diff
@@ -5,7 +5,7 @@
 import webtest
 from kinto.core import utils
-from kinto.tests.core import support as cliquet_support
+from kinto.tests.core import support as core_support
```
imports?

almet commented on kinto/tests/core/test_logging.py:
```diff
@@ -9,14 +9,14 @@
 from kinto.core import DEFAULT_SETTINGS
 from kinto.core import initialization
-from kinto.core import logs as cliquet_logs
+from kinto.core import logs as core_logs
```
ditto

almet commented on kinto/__main__.py:
```diff
@@ -4,7 +4,7 @@
 import sys
 
```
I agree that we don't really need the "cliquet" script here and that the kinto script will be enough for us.

almet commented on CHANGELOG.rst:
```diff
@@ -10,6 +10,11 @@ This document describes changes between each past release.
+  variations as part of its URL format (e.g. ``/v0``,
+  ``/v1``). Services based on kinto.core are free to use
+  ``http_api_version`` to indicate any additional changes to their
+  APIs.
```
I'm not sure what to think about that, tbh. I know we had already lenghty discussions about keeping the distinction between the protocol version defined and the actual version of kinto being used. IIRC, we wanted to keep them separate.

Keeping the protocol separated is a way to have some level of compatibility with kinto.core specific clients written in other languages, for instance. We wanted to have this to be future proof and in order to have a community around us, but I believe we failed at it so far.

As such, I wouldn't be too sad if we decide to remove it. What do @leplatrem and @natim think about this move?

Natim replied:
I think we should keep the http_protocol_version and remove the cliquet_protocol one

almet commented on docs/community.rst:
```diff
@@ -221,7 +221,7 @@ Step 1
 - Merge remaining pull requests
 - Update ``CHANGELOG.rst``
-- If protocol was updated (via *Cliquet* for example), update API changelog in
+- If protocol was updated (via *Kinto-Core* for example), update API changelog in
```
I don't think this sentence makes sense anymore ?

glasserc committed:
remove reference to kinto-core here

This doesn't make sense any more.

Natim commented:
Thanks everyone for your work on this.

I propose the following steps:
- [x] Merge this PR into kinto master
- [x] Merge the current cliquet master into kinto master (with last bugfix PR)
- [x] Add a warning on Cliquet telling that it is now a deprecated project (we might have bug fix releases for the 3.1.x branch) (Hopefully it will be easy to backport because both projects share a same git tree)
- [ ] Make a pass on the documentation
- [ ] Release a first version of the new Kinto 3.0 tree
- [ ] Make sure all plugins are up-to-date.

Does it works for you?
